src/devices/SocketCan/Interop.cs

```diff
@@ -0,0 +1,235 @@
+        {
+            CloseSocket((int)fd);
+        }
+
+        public static IntPtr CreateCanRawSocket(string networkInterface)
+        {
+            const int SOCK_RAW = 3;
+            int socket = CreateNativeSocket(PF_CAN, SOCK_RAW, CanProtocol.CAN_RAW);
+
+            if (socket == -1)
+                throw new IOException("CAN socket could not be created");
+
+            BindToInterface(socket, networkInterface);
+
+            return new IntPtr(socket);
+        }
+
+        public static bool SetCanRawSocketOption<T>(SafeHandle handle, CanSocketOption optName, ReadOnlySpan<T> data)
+            where T : struct
+        {
+            return SetSocketOption(handle, SOL_CAN_RAW, optName, data);
+        }
+
+        private static unsafe bool SetSocketOption<T>(SafeHandle handle, int level, CanSocketOption optName, ReadOnlySpan<T> data)
+            where T : struct
+        {
+            int fd = (int)handle.DangerousGetHandle();
+            ReadOnlySpan<byte> buf = MemoryMarshal.AsBytes(data);
+            fixed (byte* pinned = buf)
+            {
+                return SetSocketOpt(fd, level, (int)optName, pinned, buf.Length) == 0;
+            }
+        }
+
+        private static unsafe void BindToInterface(int fd, string interfaceName)
+        {
+            int idx = GetInterfaceIndex(fd, interfaceName);
+            CanSocketAddress addr = new CanSocketAddress();
+            addr.can_family = PF_CAN;
+            addr.can_ifindex = idx;
+
+            if (-1 == BindSocket(fd, ref addr, (uint)Marshal.SizeOf<CanSocketAddress>()))
+            {
+                throw new IOException($"Cannot bind to socket to `{interfaceName}`");
+            }
+        }
+
+        private static unsafe int GetInterfaceIndex(int fd, string name)
+        {
+            const uint SIOCGIFINDEX = 0x8933;
+            const int MaxLen = ifreq.IFNAMSIZ - 1;
+
+            if (Encoding.ASCII.GetByteCount(name) >= MaxLen)
+            {
+                throw new ArgumentException($"`{name}` exceeds maximum allowed length of {MaxLen} size", nameof(name));
+            }
+
+            ifreq ifr = new ifreq();
+            fixed (char* inIntefaceName = name)
+            {
+                int written = Encoding.ASCII.GetBytes(inIntefaceName, name.Length, ifr.ifr_name, MaxLen);
+                ifr.ifr_name[written] = 0;
+            }
+
+            int ret = Ioctl3(fd, SIOCGIFINDEX, ref ifr);
+            if (ret == -1)
+            {
+                throw new IOException($"Could not get interface index for `{name}`");
+            }
+
+            return ifr.ifr_ifindex;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal unsafe struct ifreq
+        {
+            internal const int IFNAMSIZ = 16;
+            public fixed byte ifr_name[IFNAMSIZ];
+            public int ifr_ifindex;
+            private fixed byte _padding[IFNAMSIZ - sizeof(int)];
+
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
```

Comment:
For these structs with sequential layout, should we be concerned for running on x85 vs x64 processes?

Reply:
can't be x85 at all :-)

but it doesn't matter the architecture as the size of the fields inside the struct not changing between x86 and x64

Reply:
This struct is being sent directly over the wire so 32/64 bit doesn't matter